nemo/collections/asr/models/wav2vec/wav2vec_model.py

```diff
@@ -0,0 +1,401 @@
+            logit_temp=wav2vec_cfg.logit_temp,
+        )
+
+    def training_step(self, batch, batch_idx):
+        loss, feature_loss, prob_ppl_loss = self._step(batch)
+
+        self.log('learning_rate', self._optimizer.param_groups[0]['lr'])
+        self.log('loss', loss, on_epoch=True)
+        self.log('feature_loss', feature_loss, on_epoch=True, sync_dist=True)
+        self.log('prob_ppl_loss', prob_ppl_loss, on_epoch=True, sync_dist=True)
+        return {'loss': loss}
+
+    def validation_step(self, batch, batch_idx, dataloader_idx=0):
+        loss, feature_loss, prob_ppl_loss = self._step(batch)
+        self.log('val_loss', loss, prog_bar=True, on_epoch=True, sync_dist=True)
+
+    def test_step(self, batch, batch_idx, dataloader_idx=0):
+        loss, feature_loss, prob_ppl_loss = self._step(batch)
+        self.log('test_loss', loss, prog_bar=True, on_epoch=True, sync_dist=True)
+
+    def _step(self, batch):
+        audio_signal, audio_lengths, _, _ = batch
+
+        padding_mask = self._create_padding_mask(audio_lengths)
+
+        self._update_quantizer_temp()
+        logits, targets, sampled_negatives, _, features_penalty, prob_ppl_loss, _ = self(
+            source=audio_signal, padding_mask=padding_mask
+        )
+        loss, feature_loss, prob_ppl_loss = self.loss(
+            logits=logits,
+            targets=targets,
+            negatives=sampled_negatives,
+            prob_ppl_loss=prob_ppl_loss,
+            feature_loss=features_penalty,
+        )
+        return loss, feature_loss, prob_ppl_loss
+
+    @classmethod
+    def list_available_models(cls) -> Optional[PretrainedModelInfo]:
+        return None
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "source": NeuralType(('B', 'T'), AudioSignal()),
+            "padding_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+            "mask": NeuralType(elements_type=BoolType(), optional=True),
+            "features_only": NeuralType(elements_type=BoolType(), optional=True),
+        }
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "logits": NeuralType(('B', 'T', 'D'), EncodedRepresentation()),
+            "targets": NeuralType(('B', 'T', 'D'), EncodedRepresentation(), optional=True),
+            "sampled_negatives": NeuralType(('N', 'B', 'T', 'D'), EncodedRepresentation(), optional=True),
+            "padding_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+            "features_penalty": NeuralType(elements_type=LossType(), optional=True),
+            "prob_ppl_loss": NeuralType(elements_type=LossType(), optional=True),
+            "cur_codebook_temp": NeuralType(elements_type=FloatType(), optional=True),
+        }
+
+    @typecheck()
+    def forward(self, source, padding_mask=None, mask=True, features_only=False) -> tuple:
+        prob_ppl_loss, cur_codebook_temp = None, None
```

Comment:
Would you please add have some comments on the the semantics and sizes of the inputs and outputs of the model here?

Reply:
I can do so via the `input_types` or `output_types` and add comments there, would that suffice?